example_buffer.py

```diff
@@ -78,24 +82,26 @@ def update(self, new_games, samples_per_game=4):
                 continue
             elif first_new_game is None:
                 first_new_game = idx
-                print(
-                    "Found {}/{} new games".format(len(new_games) - idx, len(new_games)))
-
-            choices = [(timestamp, ex) for ex in pick_examples_from_tfrecord(
-                game, samples_per_game)]
-            self.examples.extend(choices)
+                print("Found {}/{} new games".format(
+                    len(new_games) - idx, len(new_games)))
+            self.examples.extend(self.func(game))
 
     def flush(self, path):
+        # random.shuffle on deque is O(n^2) convert to list for O(n)
+        self.examples = list(self.examples)
+        random.shuffle(self.examples)
         with timer("Writing examples to " + path):
-            random.shuffle(self.examples)
             preprocessing.write_tf_examples(
                 path, [ex[1] for ex in self.examples])
+        self.examples.clear()
```

Comment:
after flush() is called, self.examples used to be of type Deque, but is now of type List. Subsequent fill/flush operations will have degraded complexity.

Reply:
Done

Reply:
still seems to be an issue - see `self.examples = list(self.examples)` a few lines up

Reply:
oh, okay, didn't see that.